Add OvalBorder and BoxShape.oval

bernaferrari opened the pull request:
Fix https://github.com/flutter/flutter/issues/103829

Flutter supports rectangles and circles, but doesn't support ovals (circles that are not squared). I adjusted two things:

1. Add `BoxShape.oval`. There is a comment saying new shapes shouldn't go here. I believe oval could be an exception to that, since `BoxShape.oval` is basically `BoxShape.circle` without constraints. It is easier to implement oval than circle. In most places, I just call `drawOval` instead of calculating the smaller side and drawing a circle. If you disagree, I can remove it, no problem 😀.
2.  Add OvalBorder. Works like CircleBorder, but it is unconstrained. I haven't solved lerp from Oval to Circular or from Oval to others, only it to itself (which lerps do you think are useful?). Stadium lerps to Circular and RoundedRectangle lerps to Circular.

Demo:
<img width="789" alt="image" src="https://user-images.githubusercontent.com/351125/168450448-50399f97-af41-4906-9bc7-8e5ef6d73ae8.png">

```dart
import 'package:flutter/material.dart';

const Color darkBlue = Color.fromARGB(255, 18, 32, 47);

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData.dark().copyWith(
        scaffoldBackgroundColor: darkBlue,
      ),
      debugShowCheckedModeBanner: false,
      home: const Scaffold(
        body: Center(
          child: MyWidget(),
        ),
      ),
    );
  }
}

List<Widget> spaceRow(double gap, Iterable<Widget> children) => children
    .expand((item) sync* {
      yield SizedBox(width: gap);
      yield item;
    })
    .skip(1)
    .toList();

List<Widget> spaceColumn(double gap, Iterable<Widget> children) => children
    .expand((item) sync* {
      yield SizedBox(height: gap);
      yield item;
    })
    .skip(1)
    .toList();

class MyWidget extends StatelessWidget {
  const MyWidget({super.key});

  @override
  Widget build(BuildContext context) {
    const debugStrokeAlign = StrokeAlign.inside;
    return ListView(
      children: [
        const Text("box decoration"),
        Row(
          children: [
            Container(
              width: 60,
              height: 120,
              decoration: const BoxDecoration(
                shape: BoxShape.oval,
                color: Colors.blue,
              ),
            ),
            Container(
              width: 60,
              height: 120,
              decoration: const BoxDecoration(
                shape: BoxShape.oval,
                color: Colors.blue,
                image: DecorationImage(
                  image: NetworkImage(
                      "https://unsplash.it/200/300?image=1&blur=2"),
                ),
              ),
            ),
            Container(
              width: 60,
              height: 120,
              decoration: const BoxDecoration(
                shape: BoxShape.circle,
                color: Colors.blue,
              ),
            ),
            Container(
              width: 60,
              height: 120,
              decoration: const BoxDecoration(
                shape: BoxShape.circle,
                color: Colors.blue,
                image: DecorationImage(
                  image: NetworkImage(
                      "https://unsplash.it/200/300?image=1&blur=2"),
                ),
              ),
            ),
          ],
        ),
        // const SizedBox(height: 16),
        const Text("eccentricity growing"),
        Row(
          children: [
            for (double eccentricity in [0, 0.25, 0.5, 0.75, 1])
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Container(
                  width: 100,
                  height: 160,
                  clipBehavior: Clip.antiAlias,
                  decoration: ShapeDecoration(
                    color: Colors.red,
                    shape: CircleBorder(
                      eccentricity: eccentricity,
                      side: BorderSide(
                        color: Colors.white.withOpacity(0.5),
                        width: 8,
                        strokeAlign: StrokeAlign.inside,
                      ),
                    ),
                  ),
                ),
              ),
          ],
        ),
        Row(
          children: [
            for (double eccentricity in [0, 0.25, 0.5, 0.75, 1])
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Container(
                  width: 100,
                  height: 40,
                  clipBehavior: Clip.antiAlias,
                  decoration: ShapeDecoration(
                    color: Colors.red,
                    shape: CircleBorder(
                      eccentricity: eccentricity,
                      side: BorderSide(
                        color: Colors.white.withOpacity(0.5),
                        width: 8,
                        strokeAlign: StrokeAlign.inside,
                      ),
                    ),
                  ),
                ),
              ),
          ],
        ),
        const SizedBox(height: 32),
        const Text("Lerp CircleBorder not oval -> CircleBorder oval"),
        const SizedBox(height: 16),
        Row(
          children: spaceRow(32, [
            for (double eccentricity in [0, 0.25, 0.5, 0.75, 1])
              Container(
                width: 100,
                height: 40,
                decoration: ShapeDecoration(
                  color: Colors.blue,
                  shape: CircleBorder(
                    eccentricity: 0,
                    side: BorderSide(
                      color: Colors.red.withOpacity(0.5),
                      width: 5,
                      strokeAlign: debugStrokeAlign,
                    ),
                  ).lerpTo(
                    CircleBorder(
                      eccentricity: eccentricity,
                      side: BorderSide(
                        color: Colors.white.withOpacity(0.5),
                        width: 5,
                        strokeAlign: debugStrokeAlign,
                      ),
                    ),
                    eccentricity,
                  )!,
                ),
              ),
          ]),
        ),
        const SizedBox(height: 32),
        const Text("RoundedRectangle -> Circle"),
        const SizedBox(height: 16),
        const RoundedRectangleToCircle(width: 60, height: 120),
        const SizedBox(height: 16),
        const RoundedRectangleToCircle(width: 120, height: 60),
        const SizedBox(height: 64),
        const Text("Stadium -> Circle"),
        const SizedBox(height: 16),
        const StadiumToCircle(width: 60, height: 120),
        const SizedBox(height: 16),
        const StadiumToCircle(width: 120, height: 60),
      ],
    );
  }
}

class StadiumToCircle extends StatelessWidget {
  const StadiumToCircle({
    required this.width,
    required this.height,
    Key? key,
  }) : super(key: key);

  final double width;
  final double height;

  @override
  Widget build(BuildContext context) {
    const debugStrokeAlign = StrokeAlign.inside;

    final stadiumLerp = StadiumBorder(
      side: BorderSide(
        color: Colors.red.withOpacity(0.5),
        width: 5,
        strokeAlign: StrokeAlign.inside,
      ),
    );

    ShapeBorder stadiumToCircleLerp(double t) => stadiumLerp.lerpTo(
        CircleBorder(
          eccentricity: 1,
          side: BorderSide(
            color: Colors.blue.withOpacity(0.5),
            width: 5,
            strokeAlign: StrokeAlign.inside,
          ),
        ),
        t)!;

    ShapeBorder stadiumToPartialCircleLerp(double t) => stadiumLerp.lerpTo(
        CircleBorder(
          eccentricity: 0.5,
          side: BorderSide(
            color: Colors.blue.withOpacity(0.5),
            width: 5,
            strokeAlign: debugStrokeAlign,
          ),
        ),
        t)!;

    ShapeBorder stadiumToCircleLerpFullyRound(double t) => StadiumBorder(
          side: BorderSide(
            color: Colors.blue.withOpacity(0.5),
            width: 5,
            strokeAlign: debugStrokeAlign,
          ),
        ).lerpTo(
            CircleBorder(
              eccentricity: 0.0,
              side: BorderSide(
                color: Colors.purple.withOpacity(0.5),
                width: 5,
                strokeAlign: debugStrokeAlign,
              ),
            ),
            t)!;

    return Column(
      children: [
        Row(
          children: spaceRow(32, [
            for (double t in [0, 0.25, 0.5, 0.75, 1])
              Container(
                width: width,
                height: height,
                decoration: ShapeDecoration(
                  color: Colors.red.withOpacity(0.25),
                  shape: stadiumToCircleLerp(t),
                ),
              ),
            Container(
              width: width,
              height: height,
              decoration: ShapeDecoration(
                color: Colors.white.withOpacity(0.25),
                shape: CircleBorder(
                  eccentricity: 1,
                  side: BorderSide(
                    color: Colors.white.withOpacity(0.5),
                    width: 5,
                    strokeAlign: debugStrokeAlign,
                  ),
                ),
              ),
              child: const Center(child: Text('1.0')),
            ),
          ]),
        ),
        Row(
          children: spaceRow(32, [
            for (double t in [0, 0.25, 0.5, 0.75, 1])
              Container(
                width: width,
                height: height,
                decoration: ShapeDecoration(
                  color: Colors.red.withOpacity(0.25),
                  shape: stadiumToPartialCircleLerp(t),
                ),
              ),
            Container(
              width: width,
              height: height,
              decoration: ShapeDecoration(
                color: Colors.white.withOpacity(0.25),
                shape: CircleBorder(
                  eccentricity: 0.5,
                  side: BorderSide(
                    color: Colors.white.withOpacity(0.5),
                    width: 5,
                    strokeAlign: debugStrokeAlign,
                  ),
                ),
              ),
              child: const Center(child: Text('0.5')),
            ),
          ]),
        ),
        Row(
          children: spaceRow(32, [
            for (double t in [0, 0.25, 0.5, 0.75, 1])
              Container(
                width: width,
                height: height,
                decoration: ShapeDecoration(
                  color: Colors.brown.withOpacity(0.25),
                  shape: stadiumToCircleLerpFullyRound(t),
                ),
              ),
            Container(
              width: width,
              height: height,
              decoration: ShapeDecoration(
                color: Colors.white.withOpacity(0.25),
                shape: CircleBorder(
                  eccentricity: 0.0,
                  side: BorderSide(
                    color: Colors.white.withOpacity(0.5),
                    width: 5,
                    strokeAlign: debugStrokeAlign,
                  ),
                ),
              ),
              child: const Center(child: Text('0.0')),
            ),
          ]),
        ),
      ],
    );
  }
}

class RoundedRectangleToCircle extends StatelessWidget {
  const RoundedRectangleToCircle({
    required this.width,
    required this.height,
    Key? key,
  }) : super(key: key);

  final double width;
  final double height;

  @override
  Widget build(BuildContext context) {
    const debugStrokeAlign = StrokeAlign.inside;

    final roundedLerp = RoundedRectangleBorder(
      side: BorderSide(
        color: Colors.red.withOpacity(0.5),
        width: 5,
        strokeAlign: StrokeAlign.inside,
      ),
    );

    ShapeBorder roundedToOvalLerp(double t) => roundedLerp.lerpTo(
        OvalBorder(
          side: BorderSide(
            color: Colors.blue.withOpacity(0.5),
            width: 5,
            strokeAlign: StrokeAlign.inside,
          ),
        ),
        t)!;

    ShapeBorder roundedToPartialCircleLerp(double t) => roundedLerp.lerpTo(
        CircleBorder(
          eccentricity: 0.5,
          side: BorderSide(
            color: Colors.blue.withOpacity(0.5),
            width: 5,
            strokeAlign: debugStrokeAlign,
          ),
        ),
        t)!;

    ShapeBorder roundedToCircleLerpFullyRound(double t) =>
        RoundedRectangleBorder(
          side: BorderSide(
            color: Colors.blue.withOpacity(0.5),
            width: 5,
            strokeAlign: debugStrokeAlign,
          ),
        ).lerpTo(
            CircleBorder(
              eccentricity: 0.0,
              side: BorderSide(
                color: Colors.purple.withOpacity(0.5),
                width: 5,
                strokeAlign: debugStrokeAlign,
              ),
            ),
            t)!;

    return Column(
      children: [
        Row(
          children: spaceRow(32, [
            for (double t in [0, 0.25, 0.5, 0.75, 1])
              Container(
                width: width,
                height: height,
                decoration: ShapeDecoration(
                  color: Colors.red.withOpacity(0.25),
                  shape: roundedToOvalLerp(t),
                ),
              ),
            Container(
              width: width,
              height: height,
              decoration: ShapeDecoration(
                color: Colors.white.withOpacity(0.25),
                shape: CircleBorder(
                  eccentricity: 1.0,
                  side: BorderSide(
                    color: Colors.white.withOpacity(0.5),
                    width: 5,
                    strokeAlign: debugStrokeAlign,
                  ),
                ),
              ),
              child: const Center(child: Text('1.0')),
            ),
          ]),
        ),
        Row(
          children: spaceRow(32, [
            for (double t in [0, 0.25, 0.5, 0.75, 1])
              Container(
                width: width,
                height: height,
                decoration: ShapeDecoration(
                  color: Colors.red.withOpacity(0.25),
                  shape: roundedToPartialCircleLerp(t),
                ),
              ),
            Container(
              width: width,
              height: height,
              decoration: ShapeDecoration(
                color: Colors.white.withOpacity(0.25),
                shape: CircleBorder(
                  eccentricity: 0.5,
                  side: BorderSide(
                    color: Colors.white.withOpacity(0.5),
                    width: 5,
                    strokeAlign: debugStrokeAlign,
                  ),
                ),
              ),
              child: const Center(child: Text('0.5')),
            ),
          ]),
        ),
        Row(
          children: spaceRow(32, [
            for (double t in [0, 0.25, 0.5, 0.75, 1])
              Container(
                width: width,
                height: height,
                decoration: ShapeDecoration(
                  color: Colors.brown.withOpacity(0.25),
                  shape: roundedToCircleLerpFullyRound(t),
                ),
              ),
            Container(
              width: width,
              height: height,
              decoration: ShapeDecoration(
                color: Colors.white.withOpacity(0.25),
                shape: CircleBorder(
                  eccentricity: 0.0,
                  side: BorderSide(
                    color: Colors.white.withOpacity(0.5),
                    width: 5,
                    strokeAlign: debugStrokeAlign,
                  ),
                ),
              ),
              child: const Center(child: Text('0.0')),
            ),
          ]),
        ),
      ],
    );
  }
}
```

bernaferrari committed:
Add Oval

bernaferrari committed:
Fix analyzer

bernaferrari committed:
Fix analyzer (again).

bernaferrari commented:
Cc @gspencergoog. Kind of related to the PR you just reviewed 😅. This is a lot simpler.

gspencergoog commented:
I think if we're going to add a new shape, then it should support lerping to all the other shapes.  The problem is that adding a new shape is exponential in the number of lerps you need to implement, since not only does this have to lerp to, say, rectangle, rectangle needs to lerp to it. For three shapes, you need three lerp functions.  For five, you need ten. If we had 32 shapes, we'd need 496 lerp functions. (x = n(n-1)/2). We don't have to support *every* lerp, but the more the better.  I'd say at least oval->rectangle, oval->rounded rect, oval->circle, oval->stadium, and back again.

It also makes the burden of adding a new shape exponentially harder, so we should consider if this is one of the shapes we *really* want to have.  It seems reasonably likely that it is, but we should choose carefully.

gspencergoog commented:
I like that better, since it doesn't increase the cost of adding a new shape. What about making it more "lerpable": instead of a bool, make the parameter a double that goes from 0.0 to 1.0, where 0.0 is an oval, and 1.0 is a circle. The problem is the name of this parameter: circularity? rotundity? ovalness (reverse sense)? I think I'm leaning towards `circularity`.

gspencergoog commented:
Having it be a continuum will make it more useful anyhow: I can imagine wanting a semi-oval circle even when I don't want to lerp it.

bernaferrari commented:
I'll test that, probably tomorrow. There is already a property like you described used in lerping. It will be kind of weird, but let's see the result. Because we are not lerping from round to square, it is round with min bounds to round without bounds.

gspencergoog commented:
Yes, but I think the rect to circle lerp parameter isn't the same as this.  They probably need to be combined somehow, which might be complicated. Or it might just be a multiply in the right place, if you're lucky. :-)

bernaferrari commented:
It should be simple, just the end result might be kind of weird. Let's see tomorrow.

bernaferrari commented:
👀 I added the `ovalness` but I hate that name, so I also added OvalBorder extends CircleBorder with ovalness = 1.
<img width="924" alt="image" src="https://user-images.githubusercontent.com/351125/169661100-3fc42d4f-cab4-413f-a6a9-cfbc8d66e11f.png">

bernaferrari committed:
Add ovalness attribute.

Add tests.

Improve documentation.

Fix analyzer.

bernaferrari commented:
Here it goes! I'm not happy with the `ovalness` name, but I couldn't figure a better name. Maybe `fitToSquare`, a colleague has proposed `cornerStretchAmount`, `edgeRoundnessRatio` and `cornerRoundingRatio`.

It would be better if 1 = squared circle, 0 = oval, because then `(1 - ovalness)` wouldn't be necessary (you will see when you review the code).

But, it works.
I updated the sample code with a complex example lerping everything.

BirjuVachhani committed:
Oval improvements (#2)

* Fix CircleBorder.

* Fix OvalBorder and its tests.

* Refactor ProxyBox.

bernaferrari committed:
Rename ovalness to circularity.

gspencergoog commented:
I'll take a look.  What about "eccentricity" for the name instead of "ovalness"?
I'm not sure it matches the textbook definition, but it does make some sense: eccentricity of 0.0 is a circle, and 1.0 is a very "eccentric" ellipse.

"The eccentricity of an ellipse is the ratio of the distance from the center to the foci and the distance from the center to the vertices."

bernaferrari commented:
I replaced ovalness with your circularity idea.

I think 0.0 = ellipse and 1.0 = circle makes slightly more sense codewise. (I changed from ovalness to circularity to make this change).

But eccentricity could also work. I guess first check the code, then we can discover the name we want.

bernaferrari committed:
Rename circularity to eccentricity.

Now 0.0 -> 1.0 for more ovalness.

bernaferrari committed:
Fix analyzer

bernaferrari commented:
Now it is using eccentricity 👀

gspencergoog commented:
@bernaferrari are you still wanting to merge this?  I think we're really close here.

bernaferrari commented:
Yes, I just don't know what else we need.

gspencergoog commented on packages/flutter/lib/src/painting/box_decoration.dart:
```diff
@@ -379,6 +381,9 @@ class BoxDecoration extends Decoration {
         final double distance = (position - center).distance;
         return distance <= math.min(size.width, size.height) / 2.0;
+      case BoxShape.oval:
+        final RRect bounds = RRect.fromRectAndRadius(Offset.zero & size,Radius.elliptical(size.width, size.height));
```
```suggestion
        final RRect bounds = RRect.fromRectAndRadius(Offset.zero & size, Radius.elliptical(size.width, size.height));
```

gspencergoog commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -595,8 +625,9 @@ class Border extends BoxBorder {
     }());
     assert(() {
       if (shape != BoxShape.rectangle) {
```
This should probably be a switch statement instead of an "if", just in case we add more shapes in the future.

bernaferrari replied:
It got slightly weird (wish you could return a switch :( hopefully soon), but it is there, I agree with the more shapes in future.

gspencergoog commented on packages/flutter/lib/src/painting/box_decoration.dart:
```diff
@@ -126,14 +126,14 @@ class BoxDecoration extends Decoration {
   @override
   bool debugAssertIsValid() {
-    assert(shape != BoxShape.circle || borderRadius == null); // Can't have a border radius if you're a circle.
+    assert(shape == BoxShape.rectangle || borderRadius == null); // Can't have a border radius if you're a circle or oval.
```
```suggestion
    assert(shape == BoxShape.rectangle || borderRadius == null); // Can't have a border radius if you're a circle or an oval.
```

bernaferrari replied:
Fixed, but this is definitely not following the 80-column limit.
<img width="640" alt="image" src="https://user-images.githubusercontent.com/351125/178824485-70dd1a4d-0968-4d64-91df-ceae5ac3a4fc.png">

gspencergoog commented on packages/flutter/lib/src/painting/circle_border.dart:
```diff
@@ -18,16 +18,27 @@ import 'edge_insets.dart';
   /// The [side] argument must not be null.
-  const CircleBorder({ super.side }) : assert(side != null);
+  const CircleBorder({ super.side, this.eccentricity = 0 })
+      : assert(side != null && eccentricity != null && eccentricity >= 0.0 && eccentricity <= 1.0);
```
You might split these into separate asserts for readability, and so when they fire they are more useful.

You can also print the offending value in the assert message so it's more helpful. Like so:

```dart
assert(eccentricity >= 0.0, 'The eccentricity argument $eccentricity is not greater than or equal to zero.');
```

gspencergoog replied:
Yeah, it does seem kind of silly, but there are people out there still running without strict null safety, which means that the asserts still do something.

We'll eventually make a pass and remove all such asserts with a tool once it's been long enough since we turned on null safety.

bernaferrari replied:
I thought even when null-safety was off when calling code that was null-safe it would be null-safe. Ok, adding `eccentricity != null`

gspencergoog commented on packages/flutter/lib/src/painting/oval_border.dart:
```diff
@@ -0,0 +1,54 @@
+
+/// A border that fits an elliptical shape.
+///
+/// Typically used with [ShapeDecoration] to draw an oval.
+/// Instead of centering the [Border] to a square, like [CircleBorder],
+/// it fills the available space, such that it touches the edges of the box.
+/// There is no difference between `CircleBorder(eccentricity = 1.0)` and `OvalBorder()`.
+/// [OvalBorder] works as an alias for users to discover this feature.
```
You should fill and wrap this at 80 columns:
```suggestion
/// Typically used with [ShapeDecoration] to draw an oval. Instead of centering
/// the [Border] to a square, like [CircleBorder], it fills the available space,
/// such that it touches the edges of the box. There is no difference between
/// `CircleBorder(eccentricity = 1.0)` and `OvalBorder()`. [OvalBorder] works as
/// an alias for users to discover this feature.
```

gspencergoog reviewed (changes requested):
So sorry, these comments were sitting here in GitHub, I just forgot to submit them, so I thought I was still waiting for your response!

Anyhow, here are my suggestions...

bernaferrari committed:
Fix PR review.

bernaferrari commented on packages/flutter/lib/src/painting/circle_border.dart:
```diff
@@ -18,16 +18,27 @@ import 'edge_insets.dart';
 ///
+/// The [eccentricity] parameter allows the circle to be painted touching all
+/// the edges of a rectangle, becoming an oval. When applied to a
+/// squared space, [eccentricity] is ignored.
```
changed 'the circle' to 'a circle'.

gspencergoog commented on packages/flutter/lib/src/painting/circle_border.dart:
```diff
@@ -18,16 +18,31 @@ import 'edge_insets.dart';
+  /// Defines the ratio (0.0-1.0) from which the border will be drawn
+  /// to the longest side of a rectangular box, touching all the sides.
+  /// When 0.0, it draws a circle. When 1.0, it draws an oval.
+  /// This property is ignored when applied to a squared box.
```
Can you use the same wording as above?  "a squared box" doesn't make much sense to me.

gspencergoog commented on packages/flutter/lib/src/painting/oval_border.dart:
```diff
@@ -0,0 +1,53 @@
+/// such that it touches the edges of the box. There is no difference between
+/// `CircleBorder(eccentricity = 1.0)` and `OvalBorder()`. [OvalBorder] works as
+/// an alias for users to discover this feature.
+/// See also:
```
Add a blank doc comment line before this one.

bernaferrari committed:
Add eccentricity != null.

gspencergoog commented on packages/flutter_test/lib/src/matchers.dart:
```diff
@@ -1522,6 +1522,14 @@ class _RendersOnPhysicalModel extends _MatchRenderObject<RenderPhysicalShape, Re
       return false;
     }
 
+    if (
+      borderRadius == null &&
+      (shape == BoxShape.oval || shape == BoxShape.oval) &&
+      !assertOval(shapeClipper, matchState)
+    ) {
```
Formatting here is a little different from what we normally do:

```suggestion
    if (borderRadius == null &&
        (shape == BoxShape.oval || shape == BoxShape.oval) &&
        !assertOval(shapeClipper, matchState)) {
```

bernaferrari replied:
<img width="710" alt="image" src="https://user-images.githubusercontent.com/351125/178827823-be18c27c-75a9-4692-9251-2f38e86d044a.png">

yes, but that was the reason lol

gspencergoog replied:
Ahh, OK.  Well, we don't usually format it like that.

bernaferrari replied:
Fixed the others in https://github.com/flutter/flutter/pull/103833/commits/6cfd0938e0e5434bd052f784ee9953f09a0881db

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

With these last few nits, this looks good to go!

Thanks so much for the PR, and I'm sorry it took so long for me to notice that I hadn't submitted.

bernaferrari committed:
Reword a few things.

bernaferrari committed:
Fix analyzer.

bernaferrari commented:
Don't worry, I'll forgive you if you can help give an end to the round outside StrokeAlign saga before it is too late :P

bernaferrari committed:
Tiny fix.

Hixie commented:
> I think if we're going to add a new shape, then it should support lerping to all the other shapes. The problem is that adding a new shape is exponential in the number of lerps you need to implement, since not only does this have to lerp to, say, rectangle, rectangle needs to lerp to it. For three shapes, you need three lerp functions. For five, you need ten. If we had 32 shapes, we'd need 496 lerp functions. (x = n(n-1)/2).

It's actually a bit more complicated than that, because some of the lerps actually generate new (otherwise inaccessible) types, which you then need to be able to lerp to/from. For example, _CompoundBorder, _RoundedRectangleToCircleBorder, _StadiumToRoundedRectangleBorder, etc. And do we need to support lerping from a _RoundedRectangleToCircleBorder to a _StadiumToRoundedRectangleBorder? How would we even implement that without making at least one public. :-)

Hixie commented on packages/flutter/lib/src/painting/circle_border.dart:
```diff
@@ -42,58 +58,56 @@ class CircleBorder extends OutlinedBorder {
-      return CircleBorder(side: BorderSide.lerp(a.side, side, t));
+      return CircleBorder(
+        side: BorderSide.lerp(a.side, side, t),
+        eccentricity: ui.lerpDouble(a.eccentricity, eccentricity, t)!,
```
`t` can be > 1.0 or < 0.0, which will then fire the assert in the CircleBorder constructor. We should clamp the output of lerpDouble here.

bernaferrari replied:
Oh, I never knew that! Fixed!

Hixie commented on packages/flutter/lib/src/painting/circle_border.dart:
```diff
@@ -42,58 +58,56 @@ class CircleBorder extends OutlinedBorder {
-      return CircleBorder(side: BorderSide.lerp(side, b.side, t));
+      return CircleBorder(
+        side: BorderSide.lerp(side, b.side, t),
+        eccentricity: ui.lerpDouble(eccentricity, b.eccentricity, t)!,
```
same here

bernaferrari committed:
Clamp value.

bernaferrari committed:
Add spacing and .0

bernaferrari committed:
Update oval lerp string.

bernaferrari commented:
Ok, I fixed the issue and made some tiny changes:
- Added lerpTo to OvalBorder (lerpFrom/lerpTo is optional).
- Rename `RoundedRectangleBorder(..., 50% of the way to being a CircleBorder that is 20% circular)` to `20% oval` (since now 100% = totally oval, 0% = circle)
- Added that string to StadiumBorder lerp (ops).

Ready for Google Testing again.

Hixie commented:
> [BoxShape.oval] There is a comment saying new shapes shouldn't go here.

I should probably have included more information in the comment saying why:

 - Each new enum value is a lot of new code in a bunch of places, including in code we don't control (e.g. app code, that's why the Google tests are failing), so we'd need to write a migration guide and migrate a bunch of code.
 - Each new enum value requires new code in a bunch of places.
 - That new code likely won't get tree-shaken out, since I don't think the compiler can know which values are never used. So this has a permanent binary size cost even for people who don't use it.
 - In principle, we've moved to ShapeBorders rather than BoxShape for new shapes, and the existence of BoxShape is just a legacy feature at this point. Adding more shapes encourages people to use BoxShape in the future.

bernaferrari commented:
Should I remove it then? It felt redundant because it is even simpler than the current circle option that's available there.

Hixie commented:
I would probably recommend removing it, for the binary size issue. I agree that each individual bit of code that implements `BoxShape.oval` is simpler than the equivalent code for `BoxShape.circle`, but there's a lot of these, and they all add up, and they all end up being in the compiled binary even if the user doesn't ever use either BoxShape value.

(If we could remove BoxShape entirely, I would probably do that. But that would break code that uses it.)

If we don't remove BoxShape.oval then we have to go through the breaking change process:
https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

bernaferrari committed:
Remove BoxShape.oval :(

bernaferrari committed:
Remove function.

bernaferrari commented:
Removed 😥

What if we had this lint https://github.com/dart-lang/linter/issues/3411 for BoxShape.circle only so that users 'upgrade' before removing ShapeBorder entirely?

bernaferrari committed:
Fix analyzer.

Hixie commented:
I don't think that lint necessarily is the right one (see e.g. https://github.com/dart-lang/linter/issues/3411#issuecomment-1165928053) but I could see having a specific lint that looks for BoxDecorations with BoxShape.circle and suggests using a ShapeDecoration with a CircleBorder instead, yeah. That would let us move towards deprecating BoxShape.

Actually we could just deprecate BoxShape.circle with a message saying to use ShapeDecoration with a CircleBorder instead. Is there anything that wouldn't be handled by that case?

bernaferrari commented:
Everything would be handed in that case. But... 

I don't know how it would handle this. Probably somewhere in Image accepts a shape?
```dart
decoration: const BoxDecoration(
    image: DecorationImage(
        image: AssetImage(
            'people/square/sandra.png',
             package: _kGalleryAssetsPackage,
        ),
    ),
    shape: BoxShape.circle,
),
```
And this, which is basically the only BoxShape usage in Flutter. Replace that with ShapeBorder?
<img width="789" alt="image" src="https://user-images.githubusercontent.com/351125/179826510-d126cb01-e472-4b61-9637-41610e546492.png">

Hixie commented:
I think the example above just becomes:

```dart
decoration: const ShapeDecoration(
    image: DecorationImage(
        image: AssetImage(
            'people/square/sandra.png',
             package: _kGalleryAssetsPackage,
        ),
    ),
    shape: CircleBorder(),
),
```

InkHighlight is tricky. It's unrelated to decorations and just using it for its own purposes. I would probably not touch it (though I could also imagine porting the whole ink infrastructure to ShapeBorders).

Hixie commented:
(That does mean that deprecating BoxShape.circle would be more difficult since it would affect users of InkHighlight... maybe we just deprecate `shape` on `BoxDecoration`?)

bernaferrari commented:
Yeah, seems like so :(
Maybe move the BoxShape from BoxDecoration to InkHighlighter too. 
I'll open a PR, we can play with that.

bernaferrari commented:
Ok, got halfway, but tests breaking are awful 😩 basically, everywhere that expects a circle must now expect something else, right?

```
flutter/packages/flutter/test/material/switch_test.dart -p vm --plain-name 'Switch thumb color resolves in active/enabled states'
00:05 +15 -6: Switch thumb color resolves in hovered/focused states                                                               
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure was thrown running a test:
Expected: Object or closure painting:
          'a rounded rectangle with Color(0x801e88e5), RRect: RRect.fromLTRBR(13.0, 17.0, 46.0,
31.0, 7.0)'
          'a circle with Color(0x1f000000)'
          'a circle with Color(0x33000000)'
          'a circle with Color(0x24000000)'
          'a circle with Color(0x1f000000)'
          'a circle with Color(0xff000002)'
  Actual: _RenderInkFeatures:<_RenderInkFeatures#65c76>
   Which: did not match the pattern.
          It called 5 other methods on the canvas, the first of which was drawPath(Instance of
'Path', Paint(Color(0x33000000))), but did not call drawCircle() at the time where a circle with
Color(0x33000000) was expected.
          The first method that was called when the call to drawCircle() was expected,
drawPath(Instance of 'Path', Paint(Color(0x33000000))), was called with the following stack:
```

Even this is breaking:
<img width="1049" alt="image" src="https://user-images.githubusercontent.com/351125/179842184-cf4372cb-e014-451d-b918-cf6a96aacf39.png">
And that was the whole change:
<img width="457" alt="image" src="https://user-images.githubusercontent.com/351125/179842263-5ca162dd-1cdb-4091-89bb-7c1754bb363b.png">
I need to explore that better later, or might need some help with ideas on how to fix.

bernaferrari commented:
This is the issue:
```dart
  void _paintInterior(Canvas canvas) {
    if (_interiorPaint != null) {
      canvas.drawPath(_outerPath, _interiorPaint!);
    }
  }
```
Where
```dart
@override
  Path getOuterPath(Rect rect, { TextDirection? textDirection }) {
    return Path()
      ..addOval(Rect.fromCircle(
        center: rect.center,
        radius: rect.shortestSide / 2.0,
      ));
  }
```
vs BoxDecoration:
```dart
canvas.drawCircle(center, radius, paint)
```
Is it possible to somehow convert `Path..addOval()` to `drawOval` instead of `drawPath`? This would solve the tests.

Hixie commented:
@bernaferrari can you upload a PR in draft mode with your proposed changes so far and cc me on it? Before you sink too much time into it I'm curious to see the impact.

In general, optimizing the call from drawPath(getOuterPath) to drawCircle does seem like a good idea but it's been a while since I looked at this code so I don't recall how easy that would be to do.

bernaferrari commented:
Here it goes: https://github.com/flutter/flutter/pull/108052. The changes are really simple, but the code it breaks is bad (especially switch, I think it is the worst offender).

BTW, can you add the "autosubmit" label in this PR and we finish it?

Hixie reviewed (approved):
![LGTM](http://junkyard.damowmow.com/578)
